### ISSUE TITLE
mqtt errors

### ISSUE THREAD

**Issue by GrandNewbien** [cbdevnet/midimonster#98]
Hey there, 

Super big fan of MIDIMonster here. It's truly awesome software that unlocks so many doors.

I have been getting some errors
```
mqtt    Connecting instance front to host broker.mqttdashboard.com port 8000 (TLS: no, Authentication: yes, Protocol: v5)

mqtt    Instance front disconnected, reconnection queued

winmidi Incoming data on channel lpd8.ch0.note37, value 0.244094

libmmbe Failed to send: An operation was attempted on something that is not a socket.

mqtt    Failed to transmit control message for front, assuming connection failure
```

I've been trying to use my local broker as well as an online dashboard. My config looks like this:

```
[backend winmidi]
list = on
detect = on

[winmidi lpd8]
read = LPD8

[mqtt front]
host = mqtt://username:password@192.168.0.120

[map]
lpd8.ch0.note37 > front.cmnd/note37
lpd8.ch0.cc6 > front.cmnd/cc6
```

**Comment by cbdevnet** [comment]
Hi @GrandNewbien, thanks for your interest in the MIDIMonster and your bug report!

I think you might have used the wrong port for `broker.mqttdashboard.com`, port 8000 is the WebSocket API - MQTT is on the standard 1883. Using the Websocket port probably will not respond in a way the MIDIMonster is able to understand.

Can you attach a log of the error when using a local broker?

**Comment by GrandNewbien** [comment]
Hey there, here's a local error:
```
mqtt    Connecting instance front to host 192.168.0.120 port 1883 (TLS: no, Authentication: yes, Protocol: v5)
mqtt    Instance front disconnected, reconnection queued
winmidi Incoming data on channel lpd8.ch0.cc6, value 0.511811
libmmbe Failed to send: An operation was attempted on something that is not a socket.
mqtt    Failed to transmit control message for front, assuming connection failure
```

I was able to use python to transmit midi to mqtt, but I'd love to midimonster and help out anyway I can.

**Comment by cbdevnet** [comment]
Could you tell me what broker software and what version of it you're using?

It might not yet be compatible with the newer MQTTv5, please try adding `protocol = 3.1.1` to the MQTT instance configuration for a quick test

**Comment by cbdevnet** [comment]
OK, getting the same effect on Windows - I'm investigating

**Comment by cbdevnet** [comment]
It seems there is a Windows-specific bug with Client ID handling, still investigating - for a quick fix in the meantime, please try setting an explicit `clientid` in the instance configuration.

When setting e.g. `clientid = MM`, connecting from Windows works for me

**Comment by GrandNewbien** [comment]
Setting the clientid worked! I'll try and troubleshoot locally as well. I can set a default clientid and submit a PR in a few hours if that's an acceptable workaround for now.

**Comment by cbdevnet** [comment]
I'm currently trying to analyze whether the issue has further implications.

The function that generates a random client ID at instance startup is at https://github.com/cbdevnet/midimonster/blob/master/backends/mqtt.c#L120, and it basically just calls `snprintf` to fill a 24-byte buffer with a 23-byte long client ID (minimum length required to be supported by the spec).

According to the manpages, `snprintf` will write at MOST `n` bytes _including_ the terminating NULL byte. Due to the length of the generated instance ID (which is always longer than the buffer), an edge case in the `snprintf` implementation is hit, where it has to use the last character for a terminating NULL byte. This works well on Linux, but seems to fail on Windows.

**Comment by cbdevnet** [comment]
Should be fixed in 74c1dd3b84ba9b700b60e5df99d8be265c776946, please reopen if the issue persists. Thanks for reporting!

**Comment by cbdevnet** [comment]
Builds are available from the CI at https://ci.spacecdn.de/job/midimonster/job/master/19/artifact/deployment/windows/midimonster-v0.6-23-g2e7cd46-windows.zip
